### PR TITLE
Make `IdentityInfo` generic in `pallet-identity`

### LINKED ISSUE #179: `pallet-identity`: Make `identity` generic
`IdentityInfo` is currently declared in `pallet-identity`: 
https://github.com/paritytech/substrate/blob/f6928f87138c25e0f16821a19fb1feb0edf8e214/frame/identity/src/types.rs#L290-L334

It already supports `additional` identities that are free for the user to choose. However, people are still opening issues about wanting special fields for discord, github or whatever to be added. It would be much better to make the whole `IdentityInfo` generic via the `Config` trait.

So, a `IdentityInfo` associated type should be added to the `Config` trait and everywhere where currently `IdentityInfo` is being used, should be replaced by `T::IdentityInfo`. To stay backwards compatible, we should keep the old `IdentityInfo` struct and then all downstream users can just point the `Config::IdentityInfo` to `pallet_identity::IdentityInfo`.

### PR DESCRIPTION
**Pull request by georgepisaltu**
Fixes #179 

# Description

This PR makes the structure containing identity information used in `pallet-identity` generic through the pallet `Config`. Additionally, the old structure is now available in a separate module called `simple` (pending rename) and is compatible with the new interface.

Another change in this PR is that while the `additional` field in `IdentityInfo` stays for backwards compatibility reasons, the associated costs are stil present in the pallet through the `additional` function in the `IdentityInformationProvider` interface. This function is marked as deprecated as it is only a temporary solution to the backwards compatibility problem we had. In short, we could have removed the additional fields in the struct and done a migration, but we chose to wait and do it off-chain through the genesis of the system parachain. After we move the identity pallet to the parachain, additional fields will be migrated into the existing fields and the `additional` key-value store will be removed. Until that happens, this interface will provide the necessary information to properly account for the associated costs.

Additionally, this PR fixes an unrelated issue; the `IdentityField` enum used to represent the fields as bitflags couldn't store more than 8 fields, even though it was marked as `#[repr(u64)]`. This was because of the `derive` implementation of `TypeInfo`, which assumed `u8` semantics. The custom implementation of this trait in https://github.com/paritytech/polkadot-sdk/commit/0105cc0396b7a53d0b290f48b1225847f6d17321 fixes the issue.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3758568

**Comment by georgepisaltu** [comment]
Currently the changes in this PR keep the additional fields in the `IdentityInfo` struct, so no migration would be needed. However, the purpose of this PR was to make the identity information generic and, because additional fields in a `Vec` inside the `IdentityInfo` struct is an implementation detail, the costs in terms of deposit and weight associated with these additional fields are no longer accounted for in the extrinsics of this pallet. Below are the options on how to deal with the additional fields problem:

1. Add a function in the `IdentityInformationProvider` trait to somehow query for added costs besides some agreed upon standard cost. It could be bytes stored, number of "unnecessary" entries or something similar. Convert the output of that function to an amount reserved for the storage used and some extra weight for processing.

2. Use one dedicated field for a hash/image. Any information that is not considered useful by the system but is still wanted by users is currently stored in this `additional` field. While a dedicated field for this "extra" information verifiably stores a "summary",  the actual info is stored somewhere else, off-chain maybe.

3. Remove it altogether as the purpose of the pallet is to judge information publicly. It's arguable non-standardized, custom fields and their judgements:
       a. do not have the same relevance as the standardized ones
       b. are more prone to typos and errors
       c. are more difficult to interact with from a UX perspective

Options 2 and 3 will require migrations of the [`IdentityOf` map](https://github.com/paritytech/polkadot-sdk/blob/8b061a5c5d29a4ef8efa0f2919fdc5c3cfc36361/substrate/frame/identity/src/lib.rs#L162) while option 1 would not require any. However, with a transition of the identity pallet to a system parachain we can do the migration at genesis and we would not need to worry about the additional fields costs not being accounted for if we freeze any calls to the pallet so that users can't change their identities anymore (which would be required in the build up of the move). `IdentityFields` itself doesn't require a migration since it's just a `BitFlag` of a `u64` and as long as we don't remove/change existing fields in `IdentityField`, it's backwards compatible. If we're looking for the best time to get rid of additional fields in the identity information, the genesis of the system chain would be it.

In order to safely merge this PR without accounting for the additional fields costs, we must merge the freeze of the calls to the pallet extrinsics before the next release or remove them altogether.

My personal choice is option 3 because it leaves the door open to add new fields in the future (such as the hash/image field described in option 2) in a backwards compatible way.

**Comment by sam0x17** [comment]
Just a note @georgepisaltu I was planning to migrate the identity pallet as per https://github.com/paritytech/polkadot-sdk/issues/226 but just noticed your PR so happy to make a PR off of your branch here or work on a different one instead if you'd prefer

**Comment by georgepisaltu** [comment]
> Just a note @georgepisaltu I was planning to migrate the identity pallet as per #226 but just noticed your PR so happy to make a PR off of your branch here or work on a different one instead if you'd prefer

Work on this branch, the changes are kind of related. I'll rebase the branch and resolve the conflicts as soon as possible.

**Comment by kianenigma** [comment]
> Just a note @georgepisaltu I was planning to migrate the identity pallet as per #226 but just noticed your PR so happy to make a PR off of your branch here or work on a different one instead if you'd prefer

From just reading the description, the changes that you are making are pretty orthogonal to making the identity type generic so it should be save. Feel free to open a draft PR regardless of the target branch when you have something :)

**Comment by sam0x17** [comment]
@georgepisaltu I'm happy to try to rebase if you're busy with other stuff

**Comment by georgepisaltu** [comment]
> @georgepisaltu I'm happy to try to rebase if you're busy with other stuff

I pushed the merge, I'll have to move the polkadot/kusama pallet configuration and weights to a different PR in the fellowship I think. The rest of it should be pretty much the same. Let me know if it helps.

File: substrate/frame/identity/src/simple.rs
Comment (jsidorenko) [inline substrate/frame/identity/src/simple.rs:32]: What do you think about removing those `1u64 << {x}` and having a structure to look like [here](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/nfts/src/types.rs#L257)?

File: substrate/frame/identity/src/types.rs
Comment (jsidorenko) [inline substrate/frame/identity/src/types.rs:310]: You can remove the code below and use the `impl_codec_bitflags` macro from [there](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/nfts/src/macros.rs#L29)
P. S. Previously we had a different `enumflags2` version for the identity pallet and nfts, but since we synced the versions now, we could extract that macro into a common folder and re-use the code

**Comment by ggwpez** [comment]
Summing up a recent discussion:  
- Adding the `additional_fields` method either as "deprecated" or "experimental" as intermediate solution.
- Migrating a few fields like "github" or "discord" over once we have the proper fields and deleting all the other fields.
- Removing the `additional_fields` after the transition period.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3974627

File: substrate/frame/identity/src/types.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/types.rs:338]: Unfortunately it doesn't work because of the generics, but I don't think it's too ugly either way.

File: substrate/frame/identity/src/lib.rs
Comment (joepetrowski) [inline substrate/frame/identity/src/lib.rs:374]:
```suggestion
			Ok(Some(T::WeightInfo::set_identity(judgements as u32, extra_fields)).into())
```

File: substrate/frame/identity/src/types.rs
Comment (joepetrowski) [inline substrate/frame/identity/src/types.rs:260]: Fix comment

File: substrate/frame/identity/src/types.rs
Comment (joepetrowski) [inline substrate/frame/identity/src/types.rs:241]:
```suggestion
	/// Check if an account has identity info registered for some given `fields`.
	fn has_identity(&self, fields: u64) -> bool;
```

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:374]: Done.

File: substrate/frame/identity/src/types.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/types.rs:241]: Done.

File: substrate/frame/identity/src/types.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/types.rs:260]: Done.

**Review by seadanda — COMMENTED** [review]
LGTM in general, some nits to clean up the pallet in line with the changes you've made to the docs for consistency

File: substrate/frame/identity/src/types.rs
Comment (jsidorenko) [reply substrate/frame/identity/src/types.rs:337]: hmm.. is it possible to avoid that generic? I think it could work without it, but maybe I'm missing smth

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:1]: Hopefully I got all of it now.

**Review by seadanda — APPROVED** [review]

**Comment by sam0x17** [comment]
WIP of this upgraded to fungibles is in branch `sam-upgrade-identity-to-fungible`

**Comment by sam0x17** [comment]
RE: @georgepisaltu

Yeah I will make a separate branch for this and see if I can get it working. We may need to wait for some treasury stuff to be merged. Also just making a note here in the FRAME team call some others pointed me to these links as possibly relevant:
* https://github.com/paritytech/polkadot-sdk/pull/1876/files
* https://github.com/paritytech/polkadot-sdk/pull/1785/files#diff-b7d18bc97b37b117593b3a258db3a84159771c48eb03111c0a2d794ac802d1a9R1131

**Comment by georgepisaltu** [comment]
> This LGTM now. By inspection, it should not require a state migration. Would love to see a try-runtime or Chopsticks verification that everything decodes/functions properly when upgrading to a runtime on this branch.

Although the changes pass the eye test regarding the question of migration, I ran a manual test for peace of mind. I encoded instances of both storage structures affected in this PR (`IdentityOf` -> `Registration` and `Registrars` -> `RegistrarInfo`) and, as expected, they both serialize to the same bytes as before the changes. This suggests the byte representation in storage is the same and no migration is needed.

**Review by joepetrowski — APPROVED** [review]

File: substrate/frame/identity/src/simple.rs
Comment (ggwpez) [inline substrate/frame/identity/src/simple.rs:26]: This is just the legacy stuff for out chain?  
Maybe calling it `legacy` or do you plan on maintaining it after the transition?  
Otherwise people will start using it downstream and then we have to make breaking changes to it.  
Could also be guarded by the `experimental` flag.

File: substrate/frame/identity/src/types.rs
Comment (ggwpez) [inline substrate/frame/identity/src/types.rs:235]: Why is this in storage and not a static type? Does it really need the `self` pointer?

File: substrate/frame/identity/src/lib.rs
Comment (bkchr) [inline substrate/frame/identity/src/lib.rs:582]: This makes absolutely no sense. The benchmarking in this case is wrong. `MaxAdditionFields` is not changing the weight in any way (besides more pov size, but this needs to be addressed differently)

**Comment by bkchr** [comment]
@georgepisaltu please revert the pr and reopen. This here is some kind of hybrid that should never got merged.

File: substrate/frame/identity/src/lib.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/lib.rs:582]: Looking at it now, I agree that the benchmarking here is wrong because the function doesn't iterate through any additional fields. This applies to all other extrinsics too because the `additional` field of `IdentityInfo` is only constructed and read off-chain. Knowing the length of `additional` is important only for the storage deposit.

However, the change you referenced is purely a refactor. As you can see [here](https://github.com/paritytech/polkadot-sdk/blob/35eb133baab93d3f2f179df216b2cc175d7dcaf2/substrate/frame/identity/src/lib.rs#L598C48-L598C48), functionally it is the same as before this PR. I agree that we need to fix the weights though.

Comment (joepetrowski) [reply substrate/frame/identity/src/lib.rs:582]: The benchmarking was wrong before and that should obviously be addressed. But this PR does nothing to change this behavior getting into a release. Reverting it would be the same.

File: substrate/frame/identity/src/lib.rs
Comment (bkchr) [inline substrate/frame/identity/src/lib.rs:672]: Why isn't this just a `BitVec` of length len(identity_fields).

**Comment by georgepisaltu** [comment]
> @georgepisaltu please revert the pr and reopen. This here is some kind of hybrid that should never got merged.

Reason for this approach is linked [here](https://github.com/paritytech/polkadot-sdk/pull/1661#issuecomment-1759355091), but I will try to explain briefly.

When moving identity to a system parachain, we will migrate the data through an off-chain script (PR just opened) and then we will get rid of the `additional` field in what is now `IdentityInfo` and add new supported fields, `discord` and  `github`. This is a "free" migration and we prefer it over a traditional storage migration so, in order to take advantage of that, we decided to have this  "hybrid" approach until we launch the chain, then remove the unnecessary stuff like `fn additional(&self) -> usize` from the trait bound in a runtime upgrade in the parachain (no migration necessary) and just remove the pallet from the relay. To signal that it's a temporary solution to avoid the migration, we marked the function deprecated to prevent people from using it and building on it.

This was the reasoning behind the approach. If we don't want to do this, I see 2 other options:
- revert the PR altogether and wait for the move to the system parachain, then do the change in the original issue and do a traditional migration
- remove `additional` now along with the supporting functions and do a migration now before the move to the system parachain

The benchmark issue seems orthogonal to this PR and I'd be happy to open another PR to fix that.

**Comment by bkchr** [comment]
The point being that we are not the only ones consuming these crates. Other chains out there are also using this stuff. I don't see how the offchain migration would have been affected by solving the things I highlighted.

I was probably to eager to request a revert, however this stuff should be fixed now and then backported to the release branch.

File: substrate/frame/identity/src/simple.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/simple.rs:26]: I think calling it `legacy` instead of `simple` works much better. I didn't know what to name it initially. I think it's fine if people use it downstream until we have the new version merged, the one discussed below. At the time of this PR it is in fact the current `struct` used in all chains. I'm indifferent to the `experimental` flag option.

Comment (bkchr) [reply substrate/frame/identity/src/simple.rs:26]: `experimental` shouldn't be used here as this is not "experimental". I would also have gone with `legacy`, because the indetity is not really simple as it contains quite a lot of fields.

File: substrate/frame/identity/src/types.rs
Comment (georgepisaltu) [reply substrate/frame/identity/src/types.rs:235]: I think you mean `fn has_identity(&self, fields: u64) -> bool;` which does need `self` because it's an instance query, i.e. there are multiple instances of a `struct` implementing this and some might have a subset of fields enabled, some might not. The function is not asking if the type has a field, it's asking if the instance has that field set in the bit flags. See usage [here](https://github.com/paritytech/polkadot-sdk/pull/1661/files#diff-4a95a3948151a729b1fa4e91da9c72cdc6cd2d1611842fca6f177efeed34fd1aR941) and [here](https://github.com/paritytech/polkadot-sdk/blob/c86b633695299ed27053940d5ea5c5a2392964b3/substrate/frame/alliance/src/lib.rs#L140).

**Comment by bkchr** [comment]
Sorry @georgepisaltu if my messaging here was a little bit too rude.

File: substrate/frame/identity/src/lib.rs
Comment (bkchr) [reply substrate/frame/identity/src/lib.rs:672]: What I meant there is that `IdentityFields` should have been some associated type on `IdentityInfo`. The on chain logic doesn't care what this type is.

Comment (bkchr) [reply substrate/frame/identity/src/lib.rs:672]: Yeah, but this wrapper can just go into the `legacy` module as well.

**Comment by bkchr** [comment]
Thank you @georgepisaltu!

Please also include this: https://github.com/paritytech/polkadot-sdk/pull/1661#discussion_r1371703403

> * add an extrinsic to rejig deposits to account for the change above, ensuring through proper configuration that the new deposit is always lower than whatever is reserved now to avoid any problems

Not sure we need this? I mean yes some people have some more funds reserved, but they can get them back by calling `set_identity` again. Or we could have a migration later. I'm just not a big fan of these "fix up" calls lingering around in the pallets :eyes: 

But overall sounds good! Thank you again!

**Comment by joepetrowski** [comment]
> Not sure we need this? I mean yes some people have some more funds reserved, but they can get them back by calling set_identity again. Or we could have a migration later. I'm just not a big fan of these "fix up" calls lingering around in the pallets

I don't really see it as a fix up call. There have been cases of the per-byte deposit changing [across the entire chain](https://github.com/paritytech/polkadot/commit/48de7a20d95a0ba87447e77ede531dc44cc7cd4a#diff-29b00d2d3967a817d5c82c4dae0a7e8cb10da213b80495f91ed28d3c8ffbf22aR30). Of course, if it decreases, people want to reclaim their reserved funds. In this case, calling `set_identity` again removes all judgements and would require people to go through the whole process again.

I also don't think the new deposit needs to be lower. It should just align with the normal deposit logic that charges a rate per key and per byte. Even if it increases on the Relay Chain, it will decrease significantly once it's in a parachain.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v1-3-0/4614/1

**Comment by juangirini** [comment]
> Just a note @georgepisaltu I was planning to migrate the identity pallet as per #226 but just noticed your PR so happy to make a PR off of your branch here or work on a different one instead if you'd prefer

@sam0x17 have you started working on it? If so are you planning to continue working on it or it is something I can take over?

**Comment by bkchr** [comment]
@juangirini you can do this.

**Pull request merged**
